### PR TITLE
Offline support for patching using 3.0 Website .patch files

### PR DESCRIPTION
**Pull request by Cuphat**
The electron version now allows you to select a .patch file from the 3.0 web generator as a valid Patch File. Once the patch is applied, it will output a .zpf and continue with cosmetics and other outputs.

### CODE REVIEW

File: Main.py [hunk 1]
```diff
@@ -379,13 +380,16 @@ def from_patch_file(settings, window=dummy_window()):
...
     else:
-        subfile = 'P%d.zpf' % (settings.player_num)
-        if not settings.output_file:
-            output_path += 'P%d' % (settings.player_num)
-    apply_patch_file(rom, settings.patch_file, subfile)
+        subfile = None
+        if extension == 'zpfz':
+            subfile = f"P{settings.player_num}.zpf"
+            if not settings.output_file:
+                output_path += f"P{settings.player_num}.zpf"
```
Comment (dragonbane0) [inline Main.py:391]: The original version didn't add any file extension here, so is that intended to be changed?

Comment (Cuphat) [reply Main.py]: Whoops, good catch.

**Comment by Cuphat** [comment]
I ran this through the 38 challenge seeds available in 3.0 .patch files. 3 of them produced .zpf files that were way too large (6.8MB vs. ~300KB). I'll need to take a look at those to see what is actually going on.

**Comment by Cuphat** [comment]
> I ran this through the 38 challenge seeds available in 3.0 .patch files. 3 of them produced .zpf files that were way too large (6.8MB vs. ~300KB). I'll need to take a look at those to see what is actually going on.

And resolved.

**Comment by cjohnson57** [comment]
Testing this locally, patching seemingly gets stuck at 80% and upon further inspection it seems to actually just be really slow. The loop from line 49 to 56 took over 2 minutes for my PC to run once with a length of 9 million. `new_size` in this case is 67 million, so probably related to the 6.8MB issue? I tried several of the .patch files and it happened with all of them but for example # 87

**Comment by cjohnson57** [comment]
If it isn't happening on your PC let me know if I can help narrow down the issue

**Comment by Cuphat** [comment]
It has nothing to do with the 6.8MB thing - that was just the ZPF treating certain files that were moved as if they were new and including the whole thing.

The patching is just, indeed, slow in Python. I'm not sure if there is a faster way to go about it without compiling and bundling minibsdiff (which, with the diff_roms that I wrote, might not be the worst idea). That said, it outputs a ZPF for subsequent patching or for sharing.

**Comment by cjohnson57** [comment]
Extrapolating the 9 million diff taking 2 minutes and the total size being 67 million, it'd take about 15 minutes to generate. Is that right? If so, we should definitely put some kind of messing along with patching ROM saying that it will take a while

**Comment by cjohnson57** [comment]
Also if you could figure a way to progress the bar as it goes instead of being stuck at 80% that would be sick, but I'm sure that's not an easy task

**Comment by Cuphat** [comment]
It takes 49 seconds for me to patch the whole thing. That is slow, but not so slow that it'd take 15 minutes - though I should test it on some lower-spec hardware just to be sure.

```[22:15:44] Patching ROM.
[22:16:33] Compressing ROM.```

**Comment by cjohnson57** [comment]
My computer isn't lower spec :p https://pcpartpicker.com/list/NVWgCL

Feel like there's definitely something going wrong here then, but not sure whose end it's on

**Comment by cjohnson57** [comment]
Alright, my bad. The 2 minutes thing was from me running in a debugger. Running just the app took about a minute as you said. Still, I think a minute without the progress bar moving is enough to make most people think something is wrong, so I think either of the previous suggestions still stands.

**Comment by Cuphat** [comment]
I've updated this to just bundle the minibsdiff binary and use that when available. It should be a much quicker patch, though it still needs about 20 seconds to go through and find the bytes that have changed.

**Review by cjohnson57 — APPROVED** [review]

**Comment by dragonbane0** [comment]
Creating a bsdiff patch file takes quite a while, but applying one on the website in JavaScript only took around 1 second. But I assume more is happening here to make it slower (I haven't read the implementation details)

**Pull request merged**